### PR TITLE
[Ray] Integration compiled DAG off by default

### PR DESCRIPTION
**Pull request by rkooo567**
This PR adds the experimental compiled DAG API support for VLLM NCCL opimization. It improves the benchmark_latency.py or 70B * 8 A100 GPUs' throughput benchmark around 5~7%. 

The feature is off by default, and it can be enabled using an env var (VLLM_USE_RAY_COMPILED_DAG=1). There are some rough edges we are fixing within Anyscale, and once we fix them and are more confident, we can turn this on by default.

### CODE REVIEW

**Comment by rkooo567** [comment]
cc @simon-mo can you take a look at this PR?

File: vllm/engine/llm_engine.py [hunk 1]
```diff
@@ -914,6 +930,30 @@ def _run_workers(
...
+        assert self.parallel_config.worker_use_ray
+
+        # Right now, compiled DAG requires at least 1 arg. We send
+        # a dummy value for now. It will be fixed soon.
+        with InputNode() as input_data:
+            forward_dag = MultiOutputNode([
+                worker.execute_model_compiled_dag_remote.bind(input_data)
+                for worker in self.workers
+            ])
+        return forward_dag.experimental_compile()
```
Comment (simon-mo) [inline vllm/engine/llm_engine.py:959]: which version can this run on? and when will this be non experimental?

File: vllm/engine/ray_utils.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 from vllm.config import ParallelConfig
 from vllm.logger import init_logger
 from vllm.utils import is_hip, set_cuda_visible_devices, get_ip
+import pickle
```
Comment (simon-mo) [inline vllm/engine/ray_utils.py:6]: move stdlib up

File: vllm/engine/llm_engine.py [hunk 3]
```diff
@@ -914,6 +930,30 @@ def _run_workers(
...
+                finally:
+                    # Has to call end_read in order to reuse the DAG.
+                    for chan in output_channels:
+                        chan.end_read()
+            else:
+                ray_worker_outputs = ray.get(ray_worker_outputs)
 
         return [driver_worker_output] + ray_worker_outputs
+
+    def _compiled_dag_init_dag(self):
```
Comment (simon-mo) [inline vllm/engine/llm_engine.py:948]:
```suggestion
    def _compiled_ray_dag(self):
```

File: vllm/engine/llm_engine.py [hunk 4]
```diff
@@ -122,6 +128,11 @@ def __init__(
         self.num_prompt_tokens: List[Tuple[float, int]] = []
         # List of (timestamp, num_tokens)
         self.num_generation_tokens: List[Tuple[float, int]] = []
+        # Only used when USE_RAY_COMPILED_DAG is enabled.
+        # Stores the DAG definition to run VLLM workers.
+        self.forward_dag = None
+        if USE_RAY_COMPILED_DAG:
+            self.forward_dag = self._compiled_dag_init_dag()
```
Comment (simon-mo) [inline vllm/engine/llm_engine.py:135]:
```suggestion
            self.forward_dag = self. _compiled_ray_dag()
```

File: vllm/engine/llm_engine.py [hunk 5]
```diff
@@ -897,11 +910,14 @@ def _run_workers(
             raise NotImplementedError(
                 "max_concurrent_workers is not supported yet.")
 
-        # Start the ray workers first.
-        ray_worker_outputs = [
-            worker.execute_method.remote(method, *args, **kwargs)
-            for worker in self.workers
-        ]
+        if use_ray_compiled_dag:
+            output_channels = self.forward_dag.execute(1)
```
Comment (simon-mo) [inline vllm/engine/llm_engine.py:914]: can this be done? looks like we just need any dummy input?

**Review by simon-mo — APPROVED** [review]

**Comment by rkooo567** [comment]
cc @simon-mo to merge!

**Pull request merged**